src/cascadia/WindowsTerminal/AppHost.cpp

```diff
@@ -78,38 +78,151 @@ AppHost::~AppHost()
     {
-        int argc = 0;
+        const auto result = _logic.SetStartupCommandline({ args });
+        const auto message = _logic.EarlyExitMessage();
+        if (!message.empty())
+        {
+            const auto displayHelp = result == 0;
+            const auto messageTitle = displayHelp ? IDS_HELP_DIALOG_TITLE : IDS_ERROR_DIALOG_TITLE;
+            const auto messageIcon = displayHelp ? MB_ICONWARNING : MB_ICONERROR;
+            // TODO:GH#4134: polish this dialog more, to make the text more
+            // like msiexec /?
+            MessageBoxW(nullptr,
+                        message.data(),
+                        GetStringResource(messageTitle).data(),
+                        MB_OK | messageIcon);
+
+            ExitProcess(result);
+        }
+    }
+}
+
+// Method Description:
+// - Retrieve the command line arguments passed,
+//   prepend the full name of the application in case it was missing (GH#4170),
+//   and return the arguments as vector of hstrings.
+// Arguments:
+// - <none>
+// Return Value:
+// - Program arguments as vector of hstrings. If the function fails it returns an empty vector.
+std::vector<winrt::hstring> AppHost::_NormalizedArgs() const noexcept
+{
+    try
+    {
+        // get the full name of the terminal app
+        std::wstring appPath{ wil::GetModuleFileNameW<std::wstring>(nullptr) };
 
-        // Get the argv, and turn them into a hstring array to pass to the app.
-        wil::unique_any<LPWSTR*, decltype(&::LocalFree), ::LocalFree> argv{ CommandLineToArgvW(commandline, &argc) };
-        if (argv)
+        // get the program arguments
+        std::vector<std::wstring> wstrArgs{};
+        if (!_GetArgs(wstrArgs))
         {
-            std::vector<winrt::hstring> args;
-            for (auto& elem : wil::make_range(argv.get(), argc))
+            return {};
+        }
+
+        // check if the first argument is the own call of the terminal app
+        std::vector<winrt::hstring> hstrArgs{};
+        hstrArgs.reserve(wstrArgs.size() + 1);
+        if (wstrArgs.empty())
+        {
+            hstrArgs.emplace_back(appPath);
+        }
+        else
+        {
+            // If the terminal app is in the current directory or in the
+            // PATH environment then it might be called with its base name only.
+            // So, the base name is the only part of the path we can compare.
+            // But it's even worse. The base name could be `WindowsTerminal` if
+            // called from within the IDE, it could be `wt` if the alias was called,
+            // or `wtd` for the alias of a developer's build. Thus, we only know
+            // that the base name has to have a length of at least two characters,
+            // and it has to begin with 'w' or 'W'.
+            const std::filesystem::path arg0path{ wstrArgs.front() };
+            const auto arg0name{ arg0path.stem().wstring() };
+            if (arg0name.length() < 2U || std::towlower(arg0name.front()) != L'w')

```

Comment:
So this heuristic is the only part of this PR that I don't love. I suppose that we _do_ know it's going to be in the set [`wt`, `wtd`, `WindowsTerminal`], right? Because this code would treat an argument like `whatever.exe` as the name of our application, which it most certainly _is not_. 

IMO I'd lean towards only accepting the above three, and if we find out in the future there are other possible names for us, we can add them. Thoughts?

Reply:
Honestly, I don’t see the value in stuffing something into argv 0 and validating what it is. It exists, we should ignore it, and we can copy its value to all the subcommands we execute. What that exact value is doesn’t really mean anything to us right now

Reply:
So the problem is though, sometimes arg[0] is `wt`/the application name, and we should ignore it. Other times, arg[0] is _not_ the exe name, and it's an arg that we should use. That's the problem here

Reply:
but on first launch straight off CommandlineToArgvW? That sounds a lot like somebody is calling CreateProcess badly and no application can be expected to work properly then

Reply:
I suppose that's a good point. I guess by that logic, I'm okay with 
* removing this code
* resolving #4170 as "Investigated, this only happens when the user is holding it wrong"

and merging this PR.

Reply:
So, I agree with you guys that it'll happen once in the blue moon that we either don't get any argument or that argv[0] is not the app name. Twisting CreateProcess was the only way to reproduce it. The current heuristic is the most minimalistic check I can think off. Due to the little risk I didn't want to check all three possible names in a case-insensitive way which would be quite costly compared with the benefit we gain. As long as we don't have a subcommand beginning with a `w` I assume even this minimalistic check can reduce the already low probability to a thousandth.
However, I also think it should be by far good enough to remove this piece of code entirely.

Let's see if I find a good way to compose the args in a vector of hstring without the intermediate vector of wstring. Last time I tried this the static analysis fired a ton of warnings 😆